packages/realm_dart/lib/src/logging.dart

```diff
@@ -0,0 +1,182 @@
+// Copyright 2024 MongoDB, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'dart:async';
+
+import 'package:logging/logging.dart';
+
+import 'native/realm_core.dart';
+
+// Using classes to make a fancy hierarchical enum
+sealed class RealmLogCategory {
+  /// All possible log categories.
+  static final values = [
+    realm,
+    realm.app,
+    realm.sdk,
+    realm.storage,
+    realm.storage.notification,
+    realm.storage.object,
+    realm.storage.query,
+    realm.storage.transaction,
+    realm.sync,
+    realm.sync.client,
+    realm.sync.client.changeset,
+    realm.sync.client.network,
+    realm.sync.client.reset,
+    realm.sync.client.session,
+    realm.sync.server,
+  ];
+
+  /// The root category for all log messages from the Realm SDK.
+  static final realm = _RealmLogCategory();
+
+  final String _name;
+  final RealmLogCategory? _parent;
+
+  RealmLogCategory._(this._name, this._parent);
+
+  bool contains(RealmLogCategory category) {
+    var current = category;
+    while (current != this) {
+      if (current == realm) { // root
+        return false;
+      }
+      current = current._parent!;
+    }
+    return true;
+  }
+
+  // cache the toString result
+  late final String _toString = (_parent == null ? _name : '$_parent.$_name');
+  @override
+  String toString() => _toString;
+
+  static final _map = {for (final category in RealmLogCategory.values) category.toString(): category};
+  factory RealmLogCategory.fromString(String message) => _map[message]!;
```

Comment:
Should this be public? Also, should we rename the arg to `category`?

Reply:
I'll rename argument, but I think we should keep it public.